include/deal.II/numerics/storage.h

```diff
@@ -0,0 +1,243 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_storage_h
+#define dealii_storage_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * A helper class to store a vector of shared pointers to objects of type 'T'
+ * either in FILO (First In Last Out) or FIFO (First In First Out) access
+ * scheme.
+ *
+ * The functionality is similar to std::stack but also allows random
+ * access, removal and caching.
+ *
+ * @author Denis Davydov, 2018
+ */
+template <typename T>
+class Storage
+{
+public:
+  static_assert(std::is_default_constructible<T>::value,
+                "Class T requires default-constructible elements");
+
+  /**
+   * Constructor.
+   *
+   * @parameter max_elements maximum number of elements to be stored.
+   * @parameter fifo If 'true' First In First Out access is used, otherwise
+   * First In Last Out.
+   */
+  Storage(const unsigned int max_elements = 0, const bool fifo = true);
+
+  /**
+   * Add new object by copying. If the maximum number of elements is reached,
+   * the element from the front or back will be removed based on FIFO or FILO
+   * storage scheme.
+   */
+  void
+  add(const T &element);
+
+  /**
+   * Remove an element with index @p ind.
+   */
+  void
+  remove(const unsigned int ind);
+
+  /**
+   * Read/write access to an element with index @p ind .
+   */
+  T &operator[](const unsigned int ind);
```

Comment:
How do you define the index? Relative to the first element added that's still in the list, or the last element added still in the list?